hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenBurnHandler.java

```diff
@@ -56,8 +80,98 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
         }
     }
 
+    @Override
+    public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckException {
+        final var op = txn.tokenBurnOrThrow();
+        final var fungibleCount = op.amount();
+        final var serialNums = op.serialNumbers();
+
+        validateTruePreCheck(op.hasToken(), INVALID_TOKEN_ID);
+
+        // If a positive fungible amount is present, the NFT serial numbers must be empty
+        validateFalsePreCheck(fungibleCount > 0 && !serialNums.isEmpty(), INVALID_TRANSACTION_BODY);
+
+        validateFalsePreCheck(fungibleCount < 0, INVALID_TOKEN_BURN_AMOUNT);
+
+        // Validate the NFT serial numbers
+        if (fungibleCount < 1 && !serialNums.isEmpty()) {
+            for (final var serialNumber : op.serialNumbers()) {
+                validateTruePreCheck(serialNumber > 0, INVALID_NFT_ID);
+            }
+        }
+    }
+
     @Override
     public void handle(@NonNull final HandleContext context) throws HandleException {
-        throw new UnsupportedOperationException("Not implemented");
+        requireNonNull(context);
+        final var accountStore = context.writableStore(WritableAccountStore.class);
+        final var tokenStore = context.writableStore(WritableTokenStore.class);
+        final var tokenRelStore = context.writableStore(WritableTokenRelationStore.class);
+        final var nftStore = context.writableStore(WritableNftStore.class);
+        final var txn = context.body();
+        final var op = txn.tokenBurnOrThrow();
+        final var tokenId = op.token();
+        final var fungibleBurnCount = op.amount();
+        // Wrapping the serial nums this way de-duplicates the serial nums:
+        final var nftSerialNums = new ArrayList<>(new LinkedHashSet<>(op.serialNumbers()));
+        final var validated = validateSemantics(tokenId, fungibleBurnCount, nftSerialNums, tokenStore, tokenRelStore);
+        final var token = validated.token();
+
+        if (token.tokenType() == TokenType.FUNGIBLE_COMMON) {
+            changeSupply(
+                    validated.token(),
+                    validated.tokenTreasuryRel(),
+                    -fungibleBurnCount,
+                    INVALID_TOKEN_BURN_AMOUNT,
+                    accountStore,
+                    tokenStore,
+                    tokenRelStore);
+        } else {
+            // Load and validate the nfts
```

Comment:
Should we throw `INVALID_BURN_METADATA` if we have `serialNums.esEmpty()` here

Reply:
Sure. Prehandle makes sure that doesn't happen, but I'll add it anyway

Reply:
Done

Reply:
Doesn't seem like we have this check in preHandle 🤔

Reply:
Slight restatement: `pureChecks` makes sure it doesn't. Anyway, I put it in here too

Reply:
I think we just validated if both are present and if fungible count < 0. Only if token is NonFungible we need to validate serial numbers is not empty.